walk-in-interview/src/main/java/com/google/job/servlets/JobServlet.java

```diff
@@ -0,0 +1,145 @@
+package com.google.job.servlets;
+
+import com.google.job.data.*;
+import com.google.utils.ServletUtils;
+import org.apache.commons.lang3.StringUtils;
+
+import javax.servlet.ServletException;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.util.concurrent.ExecutionException;
+import java.util.stream.Collectors;
+
+/** Servlet that handles posting new job posts and updating existing job posts. */
+@WebServlet("/jobs")
+public final class JobServlet extends HttpServlet {
+    private static final String PATCH = "PATCH";
+    private static final String JOB_ID_FIELD = "jobId";
+
+    private JobsDatabase jobsDatabase;
+
+    @Override
+    public void init() {
+        this.jobsDatabase = new JobsDatabase();
+    }
+
+    @Override
+    public void service(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
+        if (request.getMethod().equalsIgnoreCase(PATCH)){
+            doPatch(request, response);
+        } else {
+            super.service(request, response);
+        }
+    }
+
+    @Override
+    public void doPost(HttpServletRequest request, HttpServletResponse response) {
+        try {
+            // Gets job post from the form
+            Job rawJob = parseRawJobPost(request);
+
+            // Sets the status to be ACTIVE and validates the attributes via build().
+            Job job = rawJob.toBuilder().setJobStatus(JobStatus.ACTIVE).build();
+
+            // Stores job post into the database
+            storeJobPost(job);
+
+            // Sends the success status code in the response
+            response.setStatus(HttpServletResponse.SC_OK);
+        } catch (ExecutionException | IllegalArgumentException | ServletException | IOException e) {
+            // Sends the fail status code in the response
+            response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
+        }
+    }
+
+    /** Handles the PATCH request from client. */
+    public void doPatch(HttpServletRequest request, HttpServletResponse response) {
+        try {
+            // Gets the target job post id from the form
+            String jobId = ServletUtils.getStringParameter(request, JOB_ID_FIELD, /* defaultValue= */ "");
+
+            // Verifies if the current user can update the job post with this job id.
+            // TODO(issue/25): incorporate the account stuff into job post.
+            verifyUserCanUpdateJob(jobId);
+
+            // Gets raw job post from the form
+            Job rawJob = parseRawJobPost(request);
+
+            // Validates the attributes via build()
+            Job updatedJob = rawJob.toBuilder().build();
```

Comment:
personally i would validate these in the parseRawJobPost:
```
return rawJob.toBuilder().build(); // validate
```

if that's possible so that your parsing method only returns valid job objects.